NPE when using IEntityMessenger.messageSelfPeriodically();

Codec is null when scheduling message. But we are providing it. Tested on Galvan, we do not have entity messenger issues.

In `PassthroughMessengerService`, `codec.encodeMessage(message)` is null

```
  private PassthroughMessage makePassthroughMessage(EntityMessage message) throws MessageCodecException {
    @SuppressWarnings("unchecked")
    MessageCodec<EntityMessage, ?> codec = (MessageCodec<EntityMessage, ?>) this.entityContainer.codec;
    byte[] serializedMessage = codec.encodeMessage(message);
    // We use the invalid instance 0 since this is not a connected client.
    long clientInstanceID = 0;
    boolean shouldReplicateToPassives = true;
    PassthroughMessage passthroughMessage = PassthroughMessageCodec.createInvokeMessage(this.entityClassName, this.entityName, clientInstanceID, serializedMessage, shouldReplicateToPassives);
    return passthroughMessage;
  }
```

Stack trace:

```
Caused by: org.terracotta.exception.EntityUserException: Entity: org.terracotta.management.entity.nms.client.NmsEntity:CacheEntityFeaturesTest exception in user code: java.lang.NullPointerException: null
	at org.terracotta.exception.EntityUserException.writeReplace(EntityUserException.java:49)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at java.io.ObjectStreamClass.invokeWriteReplace(ObjectStreamClass.java:1118)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1136)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
	at org.terracotta.passthrough.PassthroughMessageCodec.serializeExceptionToArray(PassthroughMessageCodec.java:301)
	at org.terracotta.passthrough.PassthroughMessageCodec$10.populateStream(PassthroughMessageCodec.java:159)
	at org.terracotta.passthrough.PassthroughMessage.asSerializedBytes(PassthroughMessage.java:81)
	at org.terracotta.passthrough.PassthroughServerProcess$3.sendComplete(PassthroughServerProcess.java:392)
	at org.terracotta.passthrough.PassthroughServerMessageDecoder.sendCompleteResponse(PassthroughServerMessageDecoder.java:468)
	at org.terracotta.passthrough.PassthroughServerMessageDecoder.decode(PassthroughServerMessageDecoder.java:133)
	at org.terracotta.passthrough.PassthroughServerMessageDecoder.decode(PassthroughServerMessageDecoder.java:37)
	at org.terracotta.passthrough.PassthroughMessageCodec.runRawDecoder(PassthroughMessageCodec.java:344)
	at org.terracotta.passthrough.PassthroughMessageCodec.decodeRawMessage(PassthroughMessageCodec.java:267)
	at org.terracotta.passthrough.PassthroughServerProcess.serverThreadHandleMessage(PassthroughServerProcess.java:520)
	at org.terracotta.passthrough.PassthroughServerProcess.runServerThread(PassthroughServerProcess.java:496)
	at org.terracotta.passthrough.PassthroughServerProcess.access$000(PassthroughServerProcess.java:80)
	at org.terracotta.passthrough.PassthroughServerProcess$2.run(PassthroughServerProcess.java:262)
	... 1 more
Caused by: org.terracotta.exception.ServerSideExceptionWrapper: java.lang.NullPointerException: null
	at org.terracotta.passthrough.PassthroughMessengerService.makePassthroughMessage(PassthroughMessengerService.java:106)
	at org.terracotta.passthrough.PassthroughMessengerService.messageSelfPeriodically(PassthroughMessengerService.java:87)
	at org.terracotta.voltron.proxy.server.MessengerProxyFactory.lambda$createProxy$0(MessengerProxyFactory.java:70)
	at com.sun.proxy.$Proxy19.entityCallbackToSendMessagesToClients(Unknown Source)
	at org.terracotta.management.entity.nms.server.ActiveNmsServerEntity.createNew(ActiveNmsServerEntity.java:82)
	at org.terracotta.passthrough.PassthroughServerProcess.create(PassthroughServerProcess.java:722)
	at org.terracotta.passthrough.PassthroughServerMessageDecoder.decode(PassthroughServerMessageDecoder.java:114)
	... 8 more
```

--- a/passthrough-server/src/main/java/org/terracotta/passthrough/PassthroughCommunicatorService.java
+++ b/passthrough-server/src/main/java/org/terracotta/passthrough/PassthroughCommunicatorService.java
@@ -60,7 +60,7 @@ public class PassthroughCommunicatorService implements ClientCommunicator {
     Future<Void> waiter = connection.createClientResponseFuture();
     
     // We know that the entity better exist, by this point, to use the service.
-    CommonServerEntity<?, ?> entity = this.container.entity;
+    CommonServerEntity<?, ?> entity = this.container.getEntity();
     Assert.assertTrue(null != entity);
     byte[] payload = serialize(this.container.codec, entityMessage);
     PassthroughMessage message = PassthroughMessageCodec.createMessageToClient(clientInstanceID, payload);

--- a/passthrough-server/src/main/java/org/terracotta/passthrough/PassthroughImplementationProvidedServiceProvider.java
+++ b/passthrough-server/src/main/java/org/terracotta/passthrough/PassthroughImplementationProvidedServiceProvider.java
@@ -56,11 +56,39 @@ public interface PassthroughImplementationProvidedServiceProvider {
    * so we use this class in order to describe where an entity will eventually live when creating the service instance for
    * this entity.
    * The reason for this is that the entity constructor may ask to get a service which means that the entity isn't yet known
-   * to the platform.  So long as the entity doesn't try to use a built-in service in its constructor, but only requests it,
-   * then this allows us to defer the connection.
+   * to the platform.
+   * If the entity actually tries to use a built-in service, in its constructor, and that service needs direct access
+   * to the entity instance, it must register for notifyOnEntitySet and cache the request until it can apply it when the
+   * instance is known.
    */
   public static class DeferredEntityContainer {
-    public CommonServerEntity<?, ?> entity;
     public MessageCodec<?, ?> codec;
+    private CommonServerEntity<?, ?> entity;
+    private EntityContainerListener listener;
+    
+    public void notifyOnEntitySet(EntityContainerListener listener) {
+      // Note that our current implementation only handles one such listener (only used by PassthroughMessengerService).
+      Assert.assertTrue(null == this.listener);
+      this.listener = listener;
+    }
+    public CommonServerEntity<?, ?> getEntity() {
+      return this.entity;
+    }
+    public void setEntity(CommonServerEntity<?, ?> entity) {
+      Assert.assertTrue(null == this.entity);
+      Assert.assertTrue(null != entity);
+      this.entity = entity;
+      if (null != this.listener) {
+        this.listener.entitySetInContainer(this);
+      }
+    }
+    public void clearEntity() {
+      this.entity = null;
+    }
+  }
+
+
+  public interface EntityContainerListener {
+    public void entitySetInContainer(DeferredEntityContainer container);
   }
 }

--- a/passthrough-server/src/main/java/org/terracotta/passthrough/PassthroughMessengerService.java
+++ b/passthrough-server/src/main/java/org/terracotta/passthrough/PassthroughMessengerService.java
@@ -18,20 +18,26 @@
  */
 package org.terracotta.passthrough;
 
+import java.util.HashMap;
+import java.util.Map;
+
 import org.terracotta.entity.EntityMessage;
 import org.terracotta.entity.IEntityMessenger;
 import org.terracotta.entity.MessageCodec;
 import org.terracotta.entity.MessageCodecException;
 import org.terracotta.passthrough.PassthroughImplementationProvidedServiceProvider.DeferredEntityContainer;
+import org.terracotta.passthrough.PassthroughImplementationProvidedServiceProvider.EntityContainerListener;
 
 
-public class PassthroughMessengerService implements IEntityMessenger {
+public class PassthroughMessengerService implements IEntityMessenger, EntityContainerListener {
   private final PassthroughTimerThread timerThread;
   private final PassthroughServerProcess passthroughServerProcess;
   private final PassthroughConnection pseudoConnection;
   private final DeferredEntityContainer entityContainer;
   private final String entityClassName;
   private final String entityName;
+  // If the entity isn't yet created, we track the invocations in this map.
+  private Map<TokenWrapper, DeferredInvocation> deferredInvocations;
   
   public PassthroughMessengerService(PassthroughTimerThread timerThread, PassthroughServerProcess passthroughServerProcess, PassthroughConnection pseudoConnection, DeferredEntityContainer entityContainer, String entityClassName, String entityName) {
     this.timerThread = timerThread;
@@ -42,6 +48,13 @@ public class PassthroughMessengerService implements IEntityMessenger {
     this.entityContainer = entityContainer;
     this.entityClassName = entityClassName;
     this.entityName = entityName;
+    
+    // Since this service needs access to the entity, potentially before the entity constructor has returned, see if we
+    // need to register for notifications that the entity has been created.
+    if (null == this.entityContainer.getEntity()) {
+      this.deferredInvocations = new HashMap<TokenWrapper, DeferredInvocation>();
+      this.entityContainer.notifyOnEntitySet(this);
+    }
   }
 
   @Override
@@ -67,36 +80,80 @@ public class PassthroughMessengerService implements IEntityMessenger {
   @Override
   public ScheduledToken messageSelfAfterDelay(EntityMessage message, long millisBeforeSend) throws MessageCodecException {
     final PassthroughMessage passthroughMessage = makePassthroughMessage(message);
-    // We need to have the entity at this point.
-    Assert.assertTrue(null != this.entityContainer.entity);
-    long token = this.timerThread.scheduleAfterDelay(new Runnable(){
+    Runnable delayRunnable = new Runnable(){
       @Override
       public void run() {
         // If the entity disappeared, we are implicitly cancelled.
-        if (null != PassthroughMessengerService.this.entityContainer.entity) {
+        if (null != PassthroughMessengerService.this.entityContainer.getEntity()) {
           commonSendMessage(passthroughMessage);
         } else {
           System.err.println("WARNING:  Cancelled SINGLE delayed message to destroyed entity");
         }
-      }}, millisBeforeSend);
-    return new TokenWrapper(token);
+      }};
+    
+    TokenWrapper tokenWrapper = null;
+    // Determine if we can run this, directly, or if we need to defer this until start-up completes.
+    if (null != this.deferredInvocations) {
+      // We are still in start-up so cache this and late-bind the invocation.
+      DeferredInvocation invocation = new DeferredInvocation(delayRunnable, null, millisBeforeSend);
+      tokenWrapper = new TokenWrapper(TokenWrapper.DEFERRED);
+      this.deferredInvocations.put(tokenWrapper, invocation);
+    } else {
+      long token = this.timerThread.scheduleAfterDelay(delayRunnable, millisBeforeSend);
+      tokenWrapper = new TokenWrapper(token);
+    }
+    return tokenWrapper;
   }
 
   @Override
   public ScheduledToken messageSelfPeriodically(EntityMessage message, long millisBetweenSends) throws MessageCodecException {
     final PassthroughMessage passthroughMessage = makePassthroughMessage(message);
-    // We need to have the entity at this point.
-    Assert.assertTrue(null != this.entityContainer.entity);
     SelfCancellingRunnable runnable = new SelfCancellingRunnable(passthroughMessage);
-    long token = this.timerThread.schedulePeriodically(runnable, millisBetweenSends);
-    runnable.setToken(token);
-    return new TokenWrapper(token);
+    
+    TokenWrapper tokenWrapper = null;
+    // Determine if we can run this, directly, or if we need to defer this until start-up completes.
+    if (null != this.deferredInvocations) {
+      // We are still in start-up so cache this and late-bind the invocation.
+      DeferredInvocation invocation = new DeferredInvocation(null, runnable, millisBetweenSends);
+      tokenWrapper = new TokenWrapper(TokenWrapper.DEFERRED);
+      this.deferredInvocations.put(tokenWrapper, invocation);
+    } else {
+      long token = this.timerThread.schedulePeriodically(runnable, millisBetweenSends);
+      runnable.setToken(token);
+      tokenWrapper = new TokenWrapper(token);
+    }
+    return tokenWrapper;
   }
 
   @Override
   public void cancelTimedMessage(ScheduledToken token) {
     // If this is the wrong type, the ClassCastException is a reasonable error since it means we got something invalid.
-    this.timerThread.cancelMessage(((TokenWrapper)token).token);
+    TokenWrapper tokenWrapper = ((TokenWrapper)token);
+    if (null != this.deferredInvocations) {
+      this.deferredInvocations.remove(tokenWrapper);
+    } else {
+      this.timerThread.cancelMessage(tokenWrapper.getToken());
+    }
+  }
+
+  @Override
+  public void entitySetInContainer(DeferredEntityContainer container) {
+    for (Map.Entry<TokenWrapper, DeferredInvocation> entry : this.deferredInvocations.entrySet()) {
+      TokenWrapper wrapper = entry.getKey();
+      DeferredInvocation invocation = entry.getValue();
+      long token = -1;
+      if (null != invocation.delayRunnable) {
+        // Use the delay path.
+        token = this.timerThread.scheduleAfterDelay(invocation.delayRunnable, invocation.delayMillis);
+      } else {
+        Assert.assertTrue(null != invocation.periodicRunnable);
+        // Use the periodic path.
+        token = this.timerThread.schedulePeriodically(invocation.periodicRunnable, invocation.delayMillis);
+        invocation.periodicRunnable.setToken(token);
+      }
+      wrapper.setToken(token);
+    }
+    this.deferredInvocations = null;
   }
 
 
@@ -137,7 +194,7 @@ public class PassthroughMessengerService implements IEntityMessenger {
     public void run() {
       Assert.assertTrue(this.token > 0);
       // If the entity disappeared, we are implicitly cancelled.
-      if (null != PassthroughMessengerService.this.entityContainer.entity) {
+      if (null != PassthroughMessengerService.this.entityContainer.getEntity()) {
         commonSendMessage(this.message);
       } else {
         PassthroughMessengerService.this.timerThread.cancelMessage(this.token);
@@ -148,9 +205,31 @@ public class PassthroughMessengerService implements IEntityMessenger {
 
 
   private static class TokenWrapper implements ScheduledToken {
-    public final long token;
+    public static final long DEFERRED = -1;
+    private long token;
+    
     public TokenWrapper(long token) {
       this.token = token;
+    }
+    public long getToken() {
+      return this.token;
+    }
+    public void setToken(long token) {
+      Assert.assertTrue(DEFERRED == this.token);
+      this.token = token;
+    }
+  }
+
+
+  private static class DeferredInvocation {
+    public final Runnable delayRunnable;
+    public final SelfCancellingRunnable periodicRunnable;
+    public final long delayMillis;
+    
+    public DeferredInvocation(Runnable delayRunnable, SelfCancellingRunnable periodicRunnable, long delayMillis) {
+      this.delayRunnable = delayRunnable;
+      this.periodicRunnable = periodicRunnable;
+      this.delayMillis = delayMillis;
     }
   }
 }

--- a/passthrough-server/src/main/java/org/terracotta/passthrough/PassthroughServerProcess.java
+++ b/passthrough-server/src/main/java/org/terracotta/passthrough/PassthroughServerProcess.java
@@ -208,6 +208,10 @@ public class PassthroughServerProcess implements MessageHandler, PassthroughDump
         // We don't expect a version mismatch here or other failure in this test system.
         Assert.unexpected(e);
       }
+      // We know the codec, immediately, so pass that in.  We will need to register the entity instance after it is
+      // created.
+      container.codec = service.getMessageCodec();
+      
       PassthroughEntityTuple entityTuple = new PassthroughEntityTuple(entityData.className, entityData.entityName);
       CommonServerEntity<?, ?> newEntity = null;
       try {
@@ -217,8 +221,7 @@ public class PassthroughServerProcess implements MessageHandler, PassthroughDump
         Assert.unexpected(e);
       }
       // We can now store the entity into the deferred container.
-      container.entity = newEntity;
-      container.codec = service.getMessageCodec();
+      container.setEntity(newEntity);
       // Tell the entity to load itself from storage.
       if (newEntity instanceof ActiveServerEntity) {
         ((ActiveServerEntity<?, ?>)newEntity).loadExisting();
@@ -711,6 +714,9 @@ public class PassthroughServerProcess implements MessageHandler, PassthroughDump
     EntityServerService<?, ?> service = getServerEntityServiceForVersion(entityClassName, entityName, version);
     // This is an entity consumer so we use the deferred container.
     DeferredEntityContainer container = new DeferredEntityContainer();
+    // We know the codec, immediately, so pass that in.  We will need to register the entity instance after it is
+    // created.
+    container.codec = service.getMessageCodec();
     this.consumerToLiveContainerMap.put(consumerID, container);
     PassthroughServiceRegistry registry = getNextServiceRegistry(entityClassName, entityName, container);
     // Before we create the entity, we want to store this information regarding class and name, since that might be needed by a service in its start up.
@@ -730,8 +736,7 @@ public class PassthroughServerProcess implements MessageHandler, PassthroughDump
       // Wrap this and re-throw.
       throw new EntityConfigurationException(entityClassName, entityName, e);
     }
-    container.entity = newEntity;
-    container.codec = service.getMessageCodec();
+    container.setEntity(newEntity);
     
     // Store the tuple for this entity, so it can see itself via monitoring.
     if (null != this.serviceInterface) {
@@ -802,7 +807,7 @@ public class PassthroughServerProcess implements MessageHandler, PassthroughDump
         } catch (IOException e) {
           Assert.unexpected(e);
         }
-        container.entity = null;
+        container.clearEntity();
       }
       if (success && null != this.activeEntities) {
         boolean didRemove = false;

